packages/hierarchies-react/src/presentation-hierarchies-react/itwinui/TreeNodeRenderer.tsx

```diff
@@ -77,79 +77,94 @@ export const TreeNodeRenderer: React.ForwardRefExoticComponent<TreeNodeRendererP
-          </DropdownMenu.Root>
-        </>
-      );
+      return [
+        <TreeActionButton key={0} {...actions[0](node)} />,
+        <TreeActionButton key={1} {...actions[1](node)} />,
+        <DropdownMenu.Root key={2}>
+          <DropdownMenu.Button render={<Tree.ItemAction icon={dropdownIcon} label="Tree actions dropdown" />} />
+          <DropdownMenu.Content>
+            {actions.slice(2, actions.length).map((action) => {
+              const info = action(node);
+              return (
+                <DropdownMenu.Item key={info.label} onClick={() => info.action()}>
+                  {info.label}
+                </DropdownMenu.Item>
+              );
+            })}
+          </DropdownMenu.Content>
+        </DropdownMenu.Root>,
+      ];
     };
 
     return (
-      <Tree.Item
-        {...treeItemProps}
-        ref={ref}
-        label={getLabel ? getLabel(node) : node.label}
-        selected={selected}
-        expanded={node.isExpanded || node.children === true || node.children.length > 0 ? node.isExpanded : undefined}
-        onExpandedChange={(isExpanded) => {
-          expandNode(node.id, isExpanded);
-        }}
-        onClick={(event) => {
-          !treeItemProps["aria-disabled"] && onNodeClick?.(node, !selected, event);
-        }}
-        onKeyDown={(event) => {
-          // Ignore if it is called on the element inside, e.g. checkbox or expander
-          if (!treeItemProps["aria-disabled"] && event.target === nodeRef.current) {
-            onNodeKeyDown?.(node, !selected, event);
-          }
-        }}
-        icon={getIcon ? getIcon(node) : undefined}
-        actions={<Actions />}
-      >
-        {node.isExpanded && node.children === true ? <PlaceholderNode {...treeItemProps} ref={ref} /> : children}
-      </Tree.Item>
+      <>
+        <Tree.Item
+          {...treeItemProps}
+          ref={ref}
+          aria-level={node.level}
+          aria-posinset={node.posInLevel}
+          aria-setsize={node.levelSize}
+          label={getLabel ? getLabel(node) : node.label}
+          selected={selected}
+          expanded={node.isExpanded || node.children === true || node.children.length > 0 ? node.isExpanded : undefined}
+          onExpandedChange={(isExpanded) => {
+            expandNode(node.id, isExpanded);
+          }}
+          onClick={(event) => {
+            !treeItemProps["aria-disabled"] && onNodeClick?.(node, !selected, event);
+          }}
+          onKeyDown={(event) => {
+            // Ignore if it is called on the element inside, e.g. checkbox or expander
+            if (!treeItemProps["aria-disabled"] && event.target === nodeRef.current) {
+              onNodeKeyDown?.(node, !selected, event);
+            }
+          }}
+          icon={getIcon ? getIcon(node) : undefined}
+          actions={getActions()}
+        />
+        {node.isExpanded && node.children === true && <PlaceholderNode {...treeItemProps} level={node.level + 1} ref={ref} />}
```

Comment:
I think this might cause problems with virtualization. I think we should add entry into flattened nodes array for placeholder nodes and render it as separate `Tree.Item` component instead of rendering two components from single expanded tree node without children loaded.

Reply:
Moved placeholder rendering outside tree item, added a `placholder` boolean prop to flat tree node interface.